flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBaseTest.java

```diff
@@ -0,0 +1,352 @@
+    public void testThrowErrorOnClose() throws Exception {
+        TestCassandraOutputFormat testCassandraOutputFormat = createTestCassandraOutputFormat();
+        testCassandraOutputFormat.open(1, 1);
+
+        Exception cause = new RuntimeException();
+        testCassandraOutputFormat.enqueueCompletableFuture(
+                FutureUtils.completedExceptionally(cause));
+        testCassandraOutputFormat.writeRecord("hello");
+        try {
+            testCassandraOutputFormat.close();
+            Assert.fail("Close should have thrown an exception.");
+        } catch (IOException e) {
+            ExceptionUtils.findThrowable(e, candidate -> candidate == cause).orElseThrow(() -> e);
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testThrowErrorOnWrite() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            Exception cause = new RuntimeException();
+            testCassandraOutputFormat.enqueueCompletableFuture(
+                    FutureUtils.completedExceptionally(cause));
+
+            testCassandraOutputFormat.writeRecord("hello");
+
+            try {
+                testCassandraOutputFormat.writeRecord("world");
+                // should fail because only one completableFuture was enqueued
+                Assert.fail("Sending of second value should have failed.");
+            } catch (IOException e) {
+                Assert.assertEquals(cause, e.getCause());
+                Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+            }
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testWaitForPendingUpdatesOnClose() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            testCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+
+            testCassandraOutputFormat.writeRecord("hello");
+            Assert.assertEquals(1, testCassandraOutputFormat.getAcquiredPermits());
+
+            final CountDownLatch latch = new CountDownLatch(1);
+            Thread t =
+                    new CheckedThread("Flink-CassandraOutputFormatBaseTest") {
+                        @Override
+                        public void go() throws Exception {
+                            testCassandraOutputFormat.close();
+                            latch.countDown();
+                        }
+                    };
+            t.start();
+            while (t.getState() != Thread.State.TIMED_WAITING) {
+                Thread.sleep(5);
+            }
+
+            Assert.assertEquals(1, testCassandraOutputFormat.getAcquiredPermits());
+            // start writing
+            completableFuture.complete(null);
+            latch.await();
```

Comment:
You don't need the latch; use `CheckedThread#sync` instead.

Reply:
:+1: